charmcraft/config.py

```diff
@@ -14,171 +14,282 @@
+    loc = loc.replace(".__root__", "")
+    return loc
 
-    raise CommandError(msg)
 
+def format_pydantic_error_msg(msg):
+    """Format pydantic's error message field."""
+    # Replace shorthand "str" with "string".
+    msg = msg.replace("str type expected", "string type expected")
+    return msg
 
-@format_checker.checks("url", raises=ValueError)
-def check_url(value):
-    """Check that the URL has at least scheme and net location."""
-    if isinstance(value, str):
-        url = urlparse(value)
-        if url.scheme and url.netloc:
-            return True
-    raise ValueError("must be a full URL (e.g. 'https://some.server.com')")
 
+def format_pydantic_errors(errors):
+    """Format errors.
 
-@format_checker.checks("relative_path", raises=ValueError)
-def check_relative_paths(value):
-    """Check that the received paths are all valid relative ones."""
-    if isinstance(value, str):
-        # check if it's an absolute path using POSIX's '/' (not os.path.sep, as the charm's
-        # config is independent of the platform where charmcraft is running)
-        if value and value[0] != "/":
-            return True
-    raise ValueError("must be a valid relative URL")
+    Example 1: Single error.
 
+    Bad charmcraft.yaml content:
+    - field: <some field>
+      reason: <some reason>
 
-@attr.s(kw_only=True, frozen=True)
-class CharmhubConfig:
-    """Configuration for all Charmhub related options."""
+    Example 2: Multiple errors.
 
-    api_url = attr.ib(default="https://api.charmhub.io")
-    storage_url = attr.ib(default="https://storage.snapcraftcontent.com")
+    Bad charmcraft.yaml content:
+    - field: <some field>
+      reason: <some reason>
+    - field: <some field 2>
+      reason: <some reason 2>
+    """
+    combined = ["Bad charmcraft.yaml content:"]
+    for error in errors:
+        loc = "- field: " + format_pydantic_error_location(error["loc"])
+        combined.append(loc)
 
-    @classmethod
-    def from_dict(cls, source):
-        """Build from a raw dict."""
-        return cls(**source)
+        msg = "  reason: " + format_pydantic_error_msg(error["msg"])
+        combined.append(msg)
 
+    return "\n".join(combined)
 
-class BasicPrime(tuple):
-    """Hold the list of files to include, specified under parts/bundle/prime configs.
 
-    This is a intermediate structure until we have the full Lifecycle in place.
-    """
+class Part(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Definition of part to build."""
 
-    @classmethod
-    def from_dict(cls, parts):
-        """Build from a dicts sequence."""
-        prime = parts.get("bundle", {}).get("prime", [])
-        return cls(prime)
+    prime: List[pydantic.StrictStr] = []
 
+    @pydantic.validator("prime", each_item=True)
+    def validate_relative_paths(cls, prime):
+        """Verify relative paths are used in prime."""
+        return check_relative_paths(prime)
```

Comment:
Question: instead of declaring that the type is `StrictStr` and then apply a validation, can't we just specify that the type is something like `RelativePath` or something? Of course, `RelativePath` will be something we're coding above, that will validate it's a string and a relative path...

Reply:
Done.